timer/timer.go

```diff
@@ -0,0 +1,120 @@
+package timer
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/nivista/steady/.gen/protos/common"
+	"github.com/nivista/steady/internal/.gen/protos/messaging"
+)
+
+type (
+	// Timer represents the configuration for the execution of a recurring task.
+	Timer struct {
+		executer  executer
+		scheduler scheduler
+	}
+
+	// Progress records the progress of a recurring task.
+	Progress struct {
+		// CompletedExecutions indicated the number of successfully completed executions.
+		CompletedExecutions int
+
+		// LastExecution indicates which execution was recorded last.
+		LastExecution time.Time
+	}
+
+	executer interface {
+		execute()
+	}
+
+	scheduler interface {
+		schedule(prog Progress, now time.Time) (nextFire time.Time, done bool)
+	}
+)
+
+func getCanceller() (cancelled <-chan struct{}, cancel func()) {
+	var stopped int32
+
+	ch := make(chan struct{})
+
+	cancel = func() {
+		if atomic.CompareAndSwapInt32(&stopped, 0, 1) {
+			close(ch)
+		}
+	}
+
+	return ch, cancel
+}
+
+// Run starts the timers execution.
+func (t *Timer) Run(updateProgress func(Progress), finishTimer func(), initialProgress Progress, clock clockwork.Clock) (cancel func()) {
+
+	var progress = initialProgress
+
+	cancelled, cancel := getCanceller()
```

Comment:
Instead of returning this cancel function how about just add a method on Timer called Stop. Then you can keep all logical around starting and cleanly stopping internal to Timer

Reply:
This means the timer has state. Here's why I don't want timers to have state, timers are an isolated component so just by principle I think they should be safe for concurrent access. However, in my application I never need to access timers concurrently (because Kafka feeds me a partition in a channel that I read in one goroutine). So what I did is i just made a wrapper around timers to hold the state called timerData, it is private and is tightly integrated with all the coordinator/manager stuff. What do you think of this?

Reply:
note: timerData isn't safe for concurrent access I didn't make that clear. Also I believe there's a complexity/performance cost for making something safe for concurrent access, which is why I wanted to avoid it where it wasn't necessary to my application.